src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java

```diff
@@ -0,0 +1,97 @@
+package com.thealgorithms.datastructures.graphs;
+
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.stream.IntStream;
+
+/*
+ *  The Welsh-Powell algorithm is a graph coloring algorithm
+ *  used for coloring a graph with the minimum number of colors.
+ *  https://en.wikipedia.org/wiki/Graph_coloring
+ */
+
+public final class WelshPowell {
+    private WelshPowell() {
+    }
+
+    static class Graph {
+        private HashSet<Integer>[] adjacencyLists;
+
+        private Graph(int vertices) {
+            if (vertices < 0) {
+                throw new IllegalArgumentException("Number of vertices cannot be negative");
+            }
+
+            adjacencyLists = new HashSet[vertices];
+            Arrays.setAll(adjacencyLists, i -> new HashSet<>());
+        }
+
+        private void addEdge(int nodeA, int nodeB) {
+            validateVertex(nodeA);
+            validateVertex(nodeB);
+            if (nodeA == nodeB) {
+                throw new IllegalArgumentException("Self-loops are not allowed");
+            }
+            adjacencyLists[nodeA].add(nodeB);
+            adjacencyLists[nodeB].add(nodeA);
+        }
+
+        private void validateVertex(int vertex) {
+            if (vertex < 0 || vertex >= getNumVertices()) {
+                throw new IllegalArgumentException("Vertex " + vertex + " is out of bounds");
+            }
+        }
+
+        HashSet<Integer> getAdjList(int vertex) {
+            return adjacencyLists[vertex];
+        }
+
+        int getNumVertices() {
+            return adjacencyLists.length;
+        }
+    }
+
+    public static Graph makeGraph(int numberOfVertices, int[][] listOfEdges) {
+        Graph graph = new Graph(numberOfVertices);
+        for (int[] edge : listOfEdges) {
+            if (edge.length != 2) {
+                throw new IllegalArgumentException("Edge array must have exactly two elements");
+            }
+            graph.addEdge(edge[0], edge[1]);
+        }
+        return graph;
+    }
+
+    public static int[] findColoring(Graph graph) {
+        int[] colors = initializeColors(graph.getNumVertices());
+        Integer[] sortedVertices = getSortedNodes(graph);
+        for (int vertex : sortedVertices) {
+            if (colors[vertex] == -1) {
```

Comment:
Still, there is not test in which this condition is `false`.

This is a nice puzzle how to generate such example. Good luck!

Reply:
the only thing that crossed my mind it to add a method solely for that test, that precolors a vertex but then i have to introduce a colors[] variable to the whole code in order for it to work.

Reply:
I think there is a bug in this implementation. Please notice that color of a vertex is set only here: https://github.com/TheAlgorithms/Java/blob/a5e876f6ca6e157828b918a9440b12974c33e6d9/src/main/java/com/thealgorithms/datastructures/graphs/WelshPowell.java#L72
but we also want to color each vertex, which is not connected to the current one and not yet colored.

Reply:
Regarding c71082a5f019ba1e22b0df7dca9922e9acdefea2, I think what one should do is the following (pseudo code):
```java
    public static int[] findColoring(Graph graph) {
        int[] colors = initializeColors(graph.getNumVertices());
        Integer[] sortedVertices = getSortedNodes(graph);
        for (int vertex : sortedVertices) {
            if (colors[vertex] == -1) {
                boolean[] usedColors = computeUsedColors(graph, vertex, colors);
                final ver newColor = firstUnusedColor(usedColors);
                colors[vertex] = newColor;
                for (final var tmpVertex : getVerticesNotBeingConnectedTo(vertex)) {
                    if (colors[tmpVertex] == -1) {
                        colors[tmpVertex] = newColor;
                    }
                }
            }
        }
        return colors;
    }
```

Of course the _new part_ could be a separated method: `colorUncoloredNotNeighbours()` (or something similar). Or you could introduce a method `getUncoloredVerticesNogBeingConnectedTo(int vertex)` - that would also simplify the code.

Reply:
The above it not true (the "uncolored not-neighbors of the current node" might still be connected to each other). Sorry for confusion.

Reply:
so i do I follow the pseudocode then or do I keep it as it is?

Reply:
> so i do I follow the pseudocode then or do I keep it as it is?

Keep it as it is (or even better: refactor it). I will have a detailed look in the evening.

Reply:
okay I ll try to refactor it

Reply:
So again: regarding c71082a5f019ba1e22b0df7dca9922e9acdefea2, please revert it and use the following:
```java
    public static int[] findColoring(Graph graph) {
        int[] colors = initializeColors(graph.getNumVertices());
        Integer[] sortedVertices = getSortedNodes(graph);
        for (int vertex : sortedVertices) {
            if (colors[vertex] == -1) {
                boolean[] usedColors = computeUsedColors(graph, vertex, colors);
                final var newColor = firstUnusedColor(usedColors);
                colors[vertex] = newColor;
                Arrays.stream(sortedVertices).forEach(otherVertex -> {
                    if (colors[otherVertex] == -1 && !isAdjacentToColored(graph, otherVertex, colors)) {
                        colors[otherVertex] = newColor;
                    }
                });
            }
        }
        return colors;
    }
```
where:
```java
    private static boolean isAdjacentToColored(Graph graph, int vertex, int[] colors) {
        return graph.getAdjList(vertex).stream().anyMatch(otherVertex -> colors[otherVertex] != -1);
    }
```

Reply:
okay i am changing the code now